harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -363,158 +807,94 @@ def _auto_step_lr_scheduler_per_batch(
+                batch=batch,
+                epoch_idx=epoch_idx,
+                batch_idx=batch_idx,
+            )
 
-        # Aggregate and reduce training metrics from all the training processes.
-        if self.context.distributed.size > 1 and self.context._average_training_metrics:
-            with self.prof.record_timing("average_training_metrics"):
-                per_batch_metrics = pytorch._combine_and_average_training_metrics(
-                    self.context.distributed, per_batch_metrics
-                )
-        num_inputs *= self.context.distributed.size
-        metrics = det.util.make_metrics(num_inputs, per_batch_metrics)
+            if self.context.profiler:
+                self.context.profiler.step()
 
-        # Ignore batch_metrics entirely for custom reducers; there's no guarantee that per-batch
-        # metrics are even logical for a custom reducer.
-        with self.prof.record_timing("reduce_metrics"):
-            metrics["avg_metrics"].update(
-                pytorch._convert_metrics_to_numpy(self.context.reduce_metrics(for_training=True))
+        if self._should_update_scaler():
+            # We update the scaler once after train_batch is done because the GradScaler is
+            # expected to be one-per-training-loop, with one .update() call after all .step(opt)
+            # calls for that batch are completed [1].
+            #
+            # [1] pytorch.org/docs/master/notes/amp_examples.html
+            #         #working-with-multiple-models-losses-and-optimizers
+            self.context._scaler.update()  # type: ignore
+
+        if isinstance(training_metrics, torch.Tensor):
+            training_metrics = {"loss": training_metrics}
+
+        # Step learning rate of a pytorch.LRScheduler.
+        with self.prof.record_timing("step_lr_schedulers"):
+            for lr_scheduler in self.context.lr_schedulers:
+                self._auto_step_lr_scheduler_per_batch(batch_idx, lr_scheduler)
+
+        with self.prof.record_timing("from_device"):
+            for name, metric in training_metrics.items():
+                # Convert PyTorch metric values to NumPy, so that
+                # `det.util.encode_json` handles them properly without
+                # needing a dependency on PyTorch.
+                if isinstance(metric, torch.Tensor):
+                    metric = metric.cpu().detach().numpy()
+                training_metrics[name] = metric
+
+        batch_dur = time.time() - batch_start_time
+        samples_per_second = self.trial.get_batch_length(batch) / batch_dur
+        samples_per_second *= self.context.distributed.size
+        self.prof.record_metric("samples_per_second", samples_per_second)
+
+        if not isinstance(training_metrics, Dict):
+            raise TypeError(
+                "train_batch() must return a dictionary mapping string names to Tensor metrics, "
+                "got {type(training_metrics)}"
```

Comment:
any time you print a type, you should print `type(training_metrics).__name__`, because otherwise it will render very poorly.